src/sort.py

```diff
@@ -0,0 +1,161 @@
+
+    Returns:
+        The data structures required for sorting.
+    """
+    # Ensure initial sorting is correct
+    alpha_sorted_path_infos = sorted(path_infos, key=lambda path_info: path_info.alphabetical_rank)
+    rank_sorted_index_contents = sorted(index_contents, key=lambda item: item.rank)
+
+    directories_index = {
+        path_info.local_path: idx
+        for idx, path_info in enumerate(alpha_sorted_path_infos)
+        if path_info.local_path.is_dir()
+    }
+    directories_index[docs_path] = 0
+
+    return _SortData(
+        alpha_sorted_path_infos=alpha_sorted_path_infos,
+        local_path_yielded={path_info.local_path: False for path_info in alpha_sorted_path_infos},
+        local_path_path_info={
+            path_info.local_path: path_info for path_info in alpha_sorted_path_infos
+        },
+        directories_index=directories_index,
+        items=peekable(rank_sorted_index_contents),
+        docs_path=docs_path,
+    )
+
+
+def _contents_index_iter(
+    sort_data: _SortData,
+    current_dir: Path,
+    current_hierarchy: int = 0,
+) -> typing.Iterator[types_.PathInfo]:
+    """Recursively iterates through items by their hierarchy.
+
+    Args:
+        sort_data: The input data required for the sorting.
+        current_dir: The directory being processed.
+        current_hierarchy: The hierarchy of the directory being processed.
+
+    Yields:
+        PathInfo in sorted order first by the contents index items and then by alphabetical rank.
+    """
+    while (next_item := sort_data.items.peek(None)) is not None:
+        # Advance iterator
+        item = next_item
+        # Pass default of None to guarantee StopIteration is not raised
+        next(sort_data.items, None)
+        next_item = sort_data.items.peek(None)
+
+        # Get the path info
+        item_local_path = sort_data.docs_path / item.reference_value
+        item_path_info = sort_data.local_path_path_info[item_local_path]
+        # Update the navlink title based on the contents index
+        item_path_info_dict = item_path_info._asdict()
+        item_path_info_dict["navlink_title"] = item.reference_title
+        yield types_.PathInfo(**item_path_info_dict)
+        sort_data.local_path_yielded[item_local_path] = True
+
+        # Check for directory
+        if item_path_info.local_path.is_dir():
+            yield from _contents_index_iter(
+                sort_data=sort_data,
+                current_dir=item_path_info.local_path,
+                current_hierarchy=current_hierarchy + 1,
+            )
+
+        # Check for last item in the directory
+        if next_item is None or next_item.hierarchy <= current_hierarchy:
+            # Yield all remaining items for the current directory
+            path_infos_for_dir = itertools.takewhile(
+                lambda path_info: current_dir in path_info.local_path.parents,
+                sort_data.alpha_sorted_path_infos[sort_data.directories_index[current_dir] + 1 :],
```

Comment:
The order of that list is important, so removing items from it would change this algorithm from `O(N)` to `O(N^2)` since for each yielded item, the list has to have an element removed at a specific index which is an O(N) operation if I remember my data structures correctly (all the list items after the removed item have to be moved forward by one index)

Reply:
essentially, all the operations in this loop have to be `O(1)` so that the sorting algorithm stays `O(N)`